### DISCUSSION TITLE
TeachBooks Development Setup

### DISCUSSION THREAD

**Discussion by rlanzafame** [TeachBooks/TeachBooks#49] [Q&A]
I (@rlanzafame) am new to this, so here is a description of what I have done, I would love some feedback on it. The goal is to use it to update the `## Development ...` sections of the README.

Would be great if @CasparJungbacker and @junwon1022 could comment on what their dev setups were when adding features to teachbooks, to see if there are big differences in approach.

The setup illustrated below is how Robert has his laptop setup. The steps and packages are chosen to match our GitHub Actions workflows as closely as possible, which use `pip` and `venv`. 

```
conda deactivate
..\..\python_source\312\python -m venv venv
venv\Scripts\activate
pip install -e .
pip show teachbooks
pip install <stuff for book>
```

Notes:
- assumes users have a system with Conda installed (hence `deactivate`).
- the `..\..\python_source\312\python` path is the path to Robert's Python 3.12 installation.
- using `pip install -e .` installs the package in editable mode, which is extremely useful!
- `pip show teachbooks` is a check to see if the package is installed correctly (locally).
- as of Dec 2024 our GH workflows don't use `pip install --upgrade pip`, so this dev setup does not either.

There were issues on Robert's laptop when modifying an existing environment (i didn't use `-e` the first time, oops). Also when a second `venv` was set up in the working directory. Command `pip install -e .` got stuck; turns out there was an infinite recursive loop during the automatic package detection algorithm of `pip`; a known issue with `build` instead of `build*` in the exclude section. Changing `pyproject.toml` to include this resolved the issue (not committed to repo).
```
[tool.setuptools.packages.find]
exclude = ["tests", "docs", "venv", "venv-sphinx", ".teachbooks", "build*"]
```

**Comment by CasparJungbacker** [comment]
My setup was largely the same, but I prefer to use Anaconda/Miniconda for my environments (not because it's better, but I'm used to it :-)). 

Note that
```
pip install -e .
```
only installs the packages required for _using_ the package, not for testing or building documentation.

In `pyproject.toml`, I've defined two sets of optional dependencies:
```toml
[project.optional-dependencies]
testing = [
  "pytest",
  "flaky"
]
docs = [
  "furo",
  "numpydoc",
  "sphinx-click"
]
```
for testing and documentation, respectively. You can then install those dependencies by doing:
```
pip install -e .[testing]
```
for example.

**Reply by rlanzafame** [reply]
Yeah, I was also using conda, mostly because it is more reliable and because i like the idea that it can include non-python stuff (even though i never use it). But then I realized that all of our github actions workflows are using pip and venv's so i wanted to be more consistent with that build process (I'm not sure how much the consistency is there when i am using windows, but oh well).

nice tip for the optional dependencies - i was installing those manually one at a time!

**Comment by rlanzafame** [comment]
This discussion can be closed as the dev env setup is described in the docs (although in a simple way.
Implemented in docs in #81
